File: src/python/visualizers.py
import os
import string
import json
import random
from IPython.core.display import display, HTML
import base64
from .explanations import Explanation, AdditiveExplanation

errMsg = """
<div style='color: #900; text-align: center;'>
  <b>Visualization omitted, iML Javascript library not loaded!</b><br>
  Have you run `IML.initjs()` in this notebook? If this notebook was from another
  user you must also trust this notebook (File -> Trust notebook).
</div>"""


def initjs():
    bundlePath = os.path.join(os.path.split(__file__)[0], "..", "javascript", "build", "bundle.js")
    bundleData = open(bundlePath, "r").read()
    logoPath = os.path.join(os.path.split(__file__)[0], "..", "javascript", "build", "logoSmallGray.png")
    logoData = base64.b64encode(open(logoPath, "rb").read()).decode('ascii')
    return HTML(
        "<div align='center'><img src='data:image/png;base64,{logoData}' /></div>".format(logoData=logoData) +
        "<script>{bundleData}</script>".format(bundleData=bundleData)
    )


def id_generator(size=20, chars=string.ascii_uppercase + string.digits):
    return "i"+''.join(random.choice(chars) for _ in range(size))


def visualize(e):
    assert isinstance(e, Explanation), "visualize can only display Explanation objects!"

    if isinstance(e, AdditiveExplanation):
        return AdditiveForceVisualizer(e).html()
    else:
        return SimpleListVisualizer(e).html()


class SimpleListVisualizer:
    def __init__(self, e):
        assert isinstance(e, Explanation), "SimpleListVisualizer can only visualize Explanation objects!"

        # build the json data
        features = {}
        for i in filter(lambda j: e.effects[j] != 0, range(len(e.data.groupNames))):
            features[i] = {
                "effect": e.effects[i],
                "value": e.instance.groupDisplayValues[i]
            }
        self.data = {
            "outNames": e.model.outNames,
            "baseValue": e.baseValue,
            "link": str(e.link),
            "featureNames": e.data.groupNames,
            "features": features
        }

    def html(self):
        return HTML("""
<div id='{id}'>{errMsg}</div>
 <script>
   if (window.IML) IML.ReactDom.render(
    IML.React.createElement(IML.SimpleListVisualizer, {data}),
    document.getElementById('{id}')
  );
</script>""".format(errMsg=errMsg, data=json.dumps(self.data), id=id_generator()))


class AdditiveForceVisualizer:
    def __init__(self, e):
        assert isinstance(e, AdditiveExplanation), "AdditiveForceVisualizer can only visualize AdditiveExplanation objects!"

        # build the json data
        features = {}
        for i in filter(lambda j: e.effects[j] != 0, range(len(e.data.groupNames))):
            features[i] = {
                "effect": e.effects[i],
                "value": e.instance.groupDisplayValues[i]
            }
        self.data = {
            "outNames": e.model.outNames,
            "baseValue": e.baseValue,
            "link": str(e.link),
            "featureNames": e.data.groupNames,
            "features": features
        }

    def html(self):
        return HTML("""
<div id='{id}'>{errMsg}</div>
 <script>
   if (window.IML) IML.ReactDom.render(
    IML.React.createElement(IML.AdditiveForceVisualizer, {data}),
    document.getElementById('{id}')
  );
</script>""".format(errMsg=errMsg, data=json.dumps(self.data), id=id_generator()))

File: src/julia/visualizers.jl
using JSON
using Clustering

visualize(e::Explanation) = visualize(e, SimpleListVisualizer())
visualize(e::AdditiveExplanation) = visualize(e, AdditiveForceVisualizer())
Base.show(io::IO, ::MIME"text/html", e::Explanation) = print(io, visualize(e, SimpleListVisualizer()).content)
Base.show(io::IO, ::MIME"text/html", e::AdditiveExplanation) = print(io, visualize(e, AdditiveForceVisualizer()).content)
Base.show(io::IO, ::MIME"text/html", arr::Array{AdditiveExplanation}) = print(io, visualize(arr, AdditiveForceVisualizer()).content)

errMsg = "<div style='color: #900; text-align: center;'><b>Visualization omitted, iML Javascript library not loaded!</b><br>Have you run `IML.initjs()` in this notebook? If this notebook was from another user you must also trust this notebook (File -> Trust notebook).</div>"

type SimpleListVisualizer
end
function visualize(e::Explanation, v::SimpleListVisualizer)
    data = Dict(
        "outNames" => e.model.outNames,
        "baseValue" => e.baseValue,
        "link" => convert(String, e.link),
        "featureNames" => e.data.groupNames,
        "features" => Dict(i-1 => Dict(
            "effect" => e.effects[i],
            "value" => e.instance.groupDisplayValues[i]
        ) for i in filter(j->e.effects[j] != 0, 1:length(e.data.groupNames)))
    )

    id = randstring(20)
    return HTML(
"""
<div id='$id'>$errMsg</div>
 <script>
   if (window.IML) IML.ReactDom.render(
    IML.React.createElement(IML.SimpleListVisualizer, $(json(data))),
    document.getElementById('$id')
  );
</script>
"""
    )
end

type AdditiveForceVisualizer
end
function visualize(e::AdditiveExplanation, v::AdditiveForceVisualizer)
    data = Dict(
        "outNames" => e.model.outNames,
        "baseValue" => e.baseValue,
        "outValue" => e.outValue,
        "link" => convert(String, e.link),
        "featureNames" => e.data.groupNames,
        "features" => Dict(i-1 => Dict(
            "effect" => e.effects[i],
            "value" => e.instance.groupDisplayValues[i]
        ) for i in filter(j->e.effects[j] != 0, 1:length(e.data.groupNames)))
    )

    id = randstring(20)
    return HTML(
"""
<div id='$id'>$errMsg</div>
 <script>
   if (window.IML) IML.ReactDom.render(
    IML.React.createElement(IML.AdditiveForceVisualizer, $(json(data))),
    document.getElementById('$id')
  );
</script>
"""
    )
end
function visualize(arr::Array{AdditiveExplanation}, v::AdditiveForceVisualizer)
    # # make sure all the links and
    # @assert all([e.link == arr[1].link for e in arr])
    # @assert all([e.model == arr[1].model for e in arr])
    # @assert all([e.data == arr[1].data for e in arr])

    local clustOrder
    if all([e.model.f == arr[1].model.f for e in arr])
        m = hcat([e.effects for e in arr]...)
        D = vcat([sum((m .- m[:,i]).^2, 1) for i in 1:size(m)[2]]...)
        clustOrder = hclust(D, :complete).order
    else
        error("Tried to visualize an array of explanations from different models!")
    end

    # make sure that we put the higher predictions first...just for consistency
    if sum(arr[clustOrder[1]].effects) < sum(arr[clustOrder[end]].effects)
        reverse!(clustOrder)
    end

    clustOrder = invperm(clustOrder)
    data = Dict(
        "outNames" => arr[1].model.outNames,
        "baseValue" => arr[1].baseValue,
        "link" => convert(String, arr[1].link),
        "featureNames" => arr[1].data.groupNames,
        "explanations" => [Dict(
            "outValue" => e.outValue,
            "simIndex" => clustOrder[ind],
            "features" => Dict(i-1 => Dict(
                "effect" => e.effects[i],
                "value" => e.instance.groupDisplayValues[i]
            ) for i in filter(j->e.effects[j] != 0 || e.instance.x[j] != 0, 1:length(e.data.groupNames)))
        ) for (ind, e) in enumerate(arr)]
    )

    id = randstring(20)
    return HTML(
"""
<div id='$id'>$errMsg</div>
 <script>
   if (window.IML) IML.ReactDom.render(
    IML.React.createElement(IML.AdditiveForceArrayVisualizer, $(json(data))),
    document.getElementById('$id')
  );
</script>
"""
    )
end
